plugins/incremental-ingestion-backend/src/engine/IncrementalIngestionEngine.ts

```diff
@@ -0,0 +1,286 @@
+    if (record) {
+      this.options.logger.info(
+        `incremental-engine: Ingestion record found: '${record.id}'`,
+      );
+      return {
+        ingestionId: record.id,
+        nextAction: record.next_action as 'rest' | 'ingest' | 'backoff',
+        attempts: record.attempts as number,
+        nextActionAt: record.next_action_at.valueOf() as number,
+      };
+    }
+    const result = await this.manager.createProviderIngestionRecord(
+      providerName,
+    );
+    if (result) {
+      this.options.logger.info(
+        `incremental-engine: Ingestion record created: '${result.ingestionId}'`,
+      );
+    }
+    return result;
+  }
+
+  async ingestOneBurst(id: string, signal: AbortSignal) {
+    const lastMark = await this.manager.getLastMark(id);
+
+    const cursor = lastMark ? lastMark.cursor : void 0;
+    let sequence = lastMark ? lastMark.sequence + 1 : 0;
+
+    const start = performance.now();
+    let count = 0;
+    let done = false;
+    this.options.logger.info(
+      `incremental-engine: Ingestion '${id}' burst initiated`,
+    );
+
+    await this.options.provider.around(async (context: unknown) => {
+      let next = await this.options.provider.next(context, cursor);
+      count++;
+      // eslint-disable-next-line no-constant-condition
+      while (true) {
+        done = next.done;
+        await this.mark(id, sequence, next.entities, next.done, next.cursor);
+        if (signal.aborted || next.done) {
+          break;
+        } else {
+          next = await this.options.provider.next(context, next.cursor);
+          count++;
+          sequence++;
+        }
+      }
+    });
+
+    this.options.logger.info(
+      `incremental-engine: Ingestion '${id}' burst complete. (${count} batches in ${Math.round(
+        performance.now() - start,
+      )}ms).`,
+    );
+    return done;
+  }
+
+  async mark(
+    id: string,
+    sequence: number,
+    entities: DeferredEntity[],
+    done: boolean,
+    cursor?: unknown,
+  ) {
+    this.options.logger.debug(
+      `incremental-engine: Ingestion '${id}': MARK ${
+        entities.length
+      } entities, cursor: ${JSON.stringify(cursor)}, done: ${done}`,
+    );
+    const markId = v4();
+
+    await this.manager.createMark({
+      record: {
+        id: markId,
+        ingestion_id: id,
+        cursor,
+        sequence,
+      },
+    });
+
+    if (entities.length > 0) {
+      await this.manager.createMarkEntities(markId, entities);
+    }
+
+    const added = entities.map(deferred => ({
+      ...deferred,
+      entity: {
+        ...deferred.entity,
+        metadata: {
+          ...deferred.entity.metadata,
+          annotations: {
+            ...deferred.entity.metadata.annotations,
+            [INCREMENTAL_ENTITY_PROVIDER_ANNOTATION]:
+              this.options.provider.getProviderName(),
+          },
+        },
+      },
+    }));
+
+    const removed: DeferredEntity[] = done
+      ? []
```

Comment:
Hm. Is this NOT computed on the last page? I thought it was legal to return data with the last page - e.g. when you ask for 20 items from the upstream but it returns 18 so you conclude you're done

Reply:
Fixed! Good catch.